transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,225 @@
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(buf, "buf");
+        ObjectUtil.checkNotNull(listener, "listener");
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
+            throw new IllegalStateException("buffer queue length overflow: " + readableBytes
+                    + " + " + buf.readableBytes());
+        }
+        bufAndListenerPairs.add(buf);
+        bufAndListenerPairs.add(listener);
+        readableBytes += buf.readableBytes();
+    }
+
+
+    /**
+     * Remove a {@link ByteBuf} from the queue with the specified number of bytes. Any added buffer who's bytes are
+     * fully consumed during removal will have it's promise completed when the passed aggregate {@link ChannelPromise}
+     * completes.
+     *
+     * @param bytes the maximum number of readable bytes in the returned {@link ByteBuf}, if {@code bytes} is greater
+     *              than {@link #readableBytes} then a buffer of length {@link #readableBytes} is returned.
+     * @param aggregatePromise used to aggregate the promises and listeners for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
+        if (bytes < 0) {
+            throw new IllegalArgumentException("bytes (expected >= 0): " + bytes);
+        }
+        ObjectUtil.checkNotNull(aggregatePromise, "aggregatePromise");
+        bytes = Math.min(bytes, readableBytes);
+
+        // Use isEmpty rather than readableBytes==0 as we may have a promise associated with an empty buffer.
+        if (bufAndListenerPairs.isEmpty()) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        ByteBuf toReturn = null;
+        int originalBytes = bytes;
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                break;
+            }
+            if (entry instanceof ChannelFutureListener) {
+                aggregatePromise.addListener((ChannelFutureListener) entry);
+                continue;
+            }
+            ByteBuf entryBuffer = (ByteBuf) entry;
+            if (entryBuffer.readableBytes() > bytes) {
+                // Add the buffer back to the queue as we can't consume all of it.
+                bufAndListenerPairs.addFirst(entryBuffer);
+                if (bytes == 0) {
+                    // If we can't consume any of it just exit.
+                    break;
+                } else {
+                    // Take a slice of what we can consume and retain it.
+                    entryBuffer = entryBuffer.readSlice(bytes).retain();
+                }
+            }
+            toReturn = compose(toReturn, entryBuffer);
+            bytes -= entryBuffer.readableBytes();
+            if (entry != entryBuffer) {
+                // We couldn't consume the full entry as we took a slice of it.
+                break;
```

Comment:
Just a thought ... might it be simpler to control the loop with `while(!done)` and then set `done=true` at the top of  `if (entryBuffer.readableBytes() > bytes)`